openmc/model/triso.py

```diff
@@ -153,3 +162,843 @@ def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+        center coordinates.
+
+        'mesh'/'mesh_map' is a two way dictionary used to look up which
+        particles are located within one diameter of a given lattice cell and
+        which lattice cells a given particle center is within one diameter of.
+        This is used to speed up the nearest neighbor search.
+
+        Parameters
+        ----------
+        i : int
+            Index of particle in particles array.
+
+        """
+
+        # Determine which lattice cells the particle is in and remove the
+        # particle id from those cells
+        for idx in mesh_map[i]:
+            mesh[idx].remove(i)
+        del mesh_map[i]
+
+        # Determine which lattice cells are within one diameter of particle's
+        # center and add this particle to the list of particles in those cells
+        for idx in cell_list(particles[i], diameter):
+            mesh[idx].add(i)
+            mesh_map[i].add(idx)
+
+
+    def apply_boundary_conditions(i, j):
+        """Apply reflective boundary conditions to particles i and j.
+
+        Parameters
+        ----------
+        i, j : int
+            Index of particles in particles array.
+
+        """
+
+        for k in range(3):
+            if particles[i][k] < llim[k]:
+                particles[i][k] = llim[k]
+            elif particles[i][k] > ulim[k]:
+                particles[i][k] = ulim[k]
+            if particles[j][k] < llim[k]:
+                particles[j][k] = llim[k]
+            elif particles[j][k] > ulim[k]:
+                particles[j][k] = ulim[k]
+
+
+    def repel_particles(i, j, d):
+        """Move particles p and q apart according to the following
+        transformation (accounting for reflective boundary conditions on
+        domain):
+
+            r_i^(n+1) = r_i^(n) + 1/2(d_out^(n+1) - d^(n))
+            r_j^(n+1) = r_j^(n) - 1/2(d_out^(n+1) - d^(n))
+
+        Parameters
+        ----------
+        i, j : int
+            Index of particles in particles array.
+        d : float
+            distance between centers of particles i and j.
+
+        """
+
+        # Moving each particle distance 'r' away from the other along the line
+        # joining the particle centers will ensure their final distance is equal
+        # to the outer diameter
+        r = (outer_diameter[0] - d)/2
+
+        v = (particles[i] - particles[j])/d
+        particles[i] = particles[i] + r*v
+        particles[j] = particles[j] - r*v
+
+        # Apply reflective boundary conditions
+        apply_boundary_conditions(i, j)
+
+        update_mesh(i)
+        update_mesh(j)
+
+
+    def nearest(i):
+        """Find index of nearest neighbor of particle i.
+
+        Parameters
+        ----------
+        i : int
+            Index in particles array of particle for which to find nearest
+            neighbor.
+
+        Returns
+        -------
+        int
+            Index in particles array of nearest neighbor of i
+        double
```

Comment:
Python doesn't distinguish between float and double, so the proper terminology would be just "float"